### PR TITLE
Support filtering by quarantine dates

### PR DESCRIPTION
**Pull request by MartinBernstorff**
## Notes for reviewers
Adding a module for prediction time filtering. Added as a class as I assume many of the filters will need the same changes.

## Before merging
Update test docstring:

```
E.g. for type 2 diabetes, patients are quarantined for 730 days after
    they return to the Central Denmark Region. If a patient has an event in
    that time, they were probably incident outside of the region, and their
    following prediction times should be filtered out.

Note that this function only filters the prediction times within the quarantine.
```

### CODE REVIEW

**Review by HLasse — APPROVED** [review]
Good stuff! Gapping/censoring let's go!

File: tests/test_application_modules/test_filter_prediction_times.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+)
+from psycop_feature_generation.utils_for_testing import str_to_df
+
+
+def test_filter_by_quarantine_date():
+    """Test filtering by quarantine date.
+
+    Should filter if the prediction times is within X days after the quarantine date.
+
+    Uses dataframes as input.
```
Comment (MartinBernstorff) [inline tests/test_application_modules/test_filter_prediction_times.py:15]: Update docstring:

E.g. for type 2 diabetes, patients are quarantined for 730 days after
    they return to the Central Denmark Region. If a patient has an event in
    that time, they were probably incident outside of the region, and their
    following prediction times should be filtered out.

Note that this function only filters the prediction times within the quarantine.

File: src/psycop_feature_generation/application_modules/filter_prediction_times.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        df = df.rename(columns={"timestamp_pred": "timestamp"})
+
+        return df
+
+    def filter(self):
+        """Run filters based on the provided parameters."""
+        df = self.prediction_time_df
+
+        if self.quarantine_df is not None or self.quarantine_days is not None:
+            if self.quarantine_days is None or self.quarantine_days is None:
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/application_modules/filter_prediction_times.py:75]:
```suggestion
            if all([v is None for v in (self.quarantine_days, self.quarantine_df)]):
```

File: src/psycop_feature_generation/utils_for_testing.py [hunk 3]
```diff
@@ -47,10 +49,14 @@ def str_to_df(
         convert_timestamp_to_datetime (bool): Whether to convert the timestamp column to datetime. Defaults to True.
         convert_np_nan_to_nan (bool): Whether to convert np.nan to np.nan. Defaults to True.
         convert_str_to_float (bool): Whether to convert strings to floats. Defaults to False.
+        add_pred_time_uuid (bool): Whether to add a pred_time_uuid column. Defaults to False.
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/utils_for_testing.py:52]:
```suggestion
        add_pred_time_uuid (bool): Whether to infer a pred_time_uuid column from entity_id and timestamp columns. Defaults to False.
```

File: src/psycop_feature_generation/utils_for_testing.py [hunk 4]
```diff
@@ -39,6 +39,8 @@ def str_to_df(
     convert_timestamp_to_datetime: bool = True,
     convert_np_nan_to_nan: bool = True,
     convert_str_to_float: bool = False,
+    add_pred_time_uuid: bool = False,
+    entity_id_colname: str = "entity_id",
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/utils_for_testing.py:43]:
```suggestion
    entity_id_colname: str = "entity_id",
    timestamp_col_name: str = "timestamp",
```

File: src/psycop_feature_generation/utils_for_testing.py [hunk 5]
```diff
@@ -65,6 +71,11 @@ def str_to_df(
         # Convert all str to float
         df = df.apply(pd.to_numeric, axis=0, errors="coerce")
 
+    if add_pred_time_uuid:
+        df["pred_time_uuid"] = (
+            df[entity_id_colname].astype(str) + "_" + df["timestamp"].astype(str)
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/utils_for_testing.py:76]:
```suggestion
            df[entity_id_colname].astype(str) + "_" + df[timestamp_col_name].astype(str)
```

**Review by MartinBernstorff — COMMENTED** [review]
Misc. changes, reminders for myself

File: src/psycop_feature_generation/utils_for_testing.py [hunk 6]
```diff
@@ -47,10 +49,14 @@ def str_to_df(
...
         convert_np_nan_to_nan (bool): Whether to convert np.nan to np.nan. Defaults to True.
         convert_str_to_float (bool): Whether to convert strings to floats. Defaults to False.
+        add_pred_time_uuid (bool): Whether to add a pred_time_uuid column. Defaults to False.
+        entity_id_colname (str): The name of the entity_id column. Defaults to "entity_id".
 
     Returns:
         DataFrame: A dataframe.
     """
+    # Drop anything after the last comma
+    string = string[: string.rfind(",") + 1]
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/utils_for_testing.py:59]:
```suggestion
    # Drop comments if any exist
    string = string[: string.rfind("#")]
```

File: tests/test_application_modules/test_filter_prediction_times.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
+"""Tests of the filter_prediction_times module."""
+import pandas as pd
+
+from psycop_feature_generation.application_modules.filter_prediction_times import (
+    PredictionTimeFilterer,
+)
+from psycop_feature_generation.utils_for_testing import str_to_df
+
+
+def test_filter_by_quarantine_date():
```
Comment (frillecode) [inline tests/test_application_modules/test_filter_prediction_times.py:10]: here you use "...._quarantine_date" but in the PredictionTimeFilterer you use "....._quarantine_period",  maybe use the same term across? not that important but perhaps more readable

File: tests/test_application_modules/test_filter_prediction_times.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    """Test filtering by quarantine date.
+
+    Should filter if the prediction times is within X days after the quarantine date.
```
Comment (frillecode) [inline tests/test_application_modules/test_filter_prediction_times.py:13]: i.e. lies within the quarantine period? Again, just to make it more readable by calling it the same thing across

**Review by frillecode — APPROVED** [review]
Looks good, Martin! :-) 
Just added a few suggestions for making it more readable. In general, I think more informative docstrings in the `PredictionTimeFilterer` class could solve much of this.

**Pull request merged**